### DISCUSSION TITLE
Can't migrate , Foreign Key Database Error 42804

### DISCUSSION THREAD

**Discussion by AliMejbar** [prisma/prisma#13774] [Q&A] [answered]
Hello, I'm trying to achieve a schema where a user has a field `authTokens` containing the list of all the tokens that were generated for him, for logging and also the ability to blacklist one. A many to many relationship between `User` and `Device` through `UserOnDevice` so that a user can have multiple devices and a device may have different users, and also link it to the `AuthTokens` to for example disconnect or discriminate by device and not with a token name directly. I couldn't find any help online, Thank's for your time.

Here is the part of my `schema.prisma` file that are causing the error.

```prisma
model User {
  id                        Int              @id @default(autoincrement())
  email                     String           @unique
  phone                     String           @unique
  password                  String
  userOnDevice              UserOnDevice[]
  authTokens                AuthToken[]
  // ... other fields
}
model AuthToken {
  id          String        @id @default(cuid())
  tokenType   AuthTokenType @default(ACCESS_TOKEN)
  token       String        @unique
  blackListed Boolean       @default(false)
  user        User          @relation(fields: [userId], references: [id], onDelete: Cascade)
  userId      Int
  createdAt   DateTime      @default(now())
  device      Device?       @relation(fields: [deviceId], references: [id], onDelete: SetNull)
  deviceId    String?
}

enum AuthTokenType {
  ACCESS_TOKEN
  REFRESH_TOKEN
}

model UserOnDevice {
  id        Int      @id @default(autoincrement())
  user      User     @relation(fields: [userId], references: [id], onDelete: Cascade)
  userId    Int
  device    Device   @relation(fields: [deviceId], references: [id], onDelete: Cascade)
  deviceId  String
  createdAt DateTime @default(now())

  @@unique([userId, deviceId])
}

model Device {
  id           String         @id @default(uuid()) @db.Uuid
  userOnDevice UserOnDevice[]
  platform     PLATFORM       @default(unknown)
  name         String?
  model        String?
  manufacturer String?
  osVersion    String?
  createdAt    DateTime       @default(now())
  authTokens   AuthToken[]
}
```

Terminal error output : 

![image](https://user-images.githubusercontent.com/55496459/173256751-479563d7-d667-4a13-b138-ba7ddc5245c1.png)

**Comment by janpio** [comment, marked as answer]
`Device.id` is of type `String @db.Uuid`, but the other two `deviceId` fields are just `String` and missing the `@db.Uuid`. Add that, and it should work.

What version are you using? In theory that should already be a validation error in the schema as far as I know.

**Reply by AliMejbar** [reply]
i'm using "prisma": "^3.15.1", aka latest and i have no validation error , I'll try this and get back to you in a second.

**Reply by janpio** [reply]
Correct, we even have an issue for this: https://github.com/prisma/prisma/issues/13005 Sorry, I thought we had already fixed that.

**Reply by AliMejbar** [reply]
I still get the error even with 
```prisma
model AuthToken {
  id          String        @id @default(cuid())
  tokenType   AuthTokenType @default(ACCESS_TOKEN)
  token       String        @unique
  blackListed Boolean       @default(false)
  user        User          @relation(fields: [userId], references: [id], onDelete: Cascade)
  userId      Int
  createdAt   DateTime      @default(now())
  device      Device?       @relation(fields: [deviceId], references: [id], onDelete: SetNull)
  deviceId    String?       @db.Uuid
}
model UserOnDevice {
  id        Int      @id @default(autoincrement())
  user      User     @relation(fields: [userId], references: [id], onDelete: Cascade)
  userId    Int
  device    Device   @relation(fields: [deviceId], references: [id], onDelete: Cascade)
  deviceId  String   @db.Uuid
  createdAt DateTime @default(now())

  @@unique([userId, deviceId])
}
```

![image](https://user-images.githubusercontent.com/55496459/173258766-ba8cc90a-3b5e-4578-ab65-5d3f794b50bb.png)

**Reply by janpio** [reply]
(The schema snippet you posted is missing the `Device` model, but if you did not change that it should work now.)

Did you delete the previously created migration file after you change the schema file?
If not, sharing the full schema again would be nice.

**Comment by AliMejbar** [comment]
Adding `@db.Uuid`and deleting the blocking migration fixed the problem. Thank's @janpio for your help.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
